rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -1,23 +1,63 @@
 use async_trait::async_trait;
 use hyperlane_core::{
-    BlockInfo, ChainResult, HyperlaneChain, HyperlaneDomain, HyperlaneProvider, TxnInfo, H256,
+    metrics::agent::AgentMetricsFetcher, BlockInfo, ChainResult, ContractLocator, HyperlaneChain,
+    HyperlaneDomain, HyperlaneProvider, TxnInfo, H256, U256,
 };
+use tendermint_rpc::{client::CompatMode, HttpClient};
+
+use crate::{ConnectionConf, HyperlaneCosmosError, Signer};
+
+use self::grpc::WasmGrpcProvider;
 
 /// cosmos grpc provider
 pub mod grpc;
 /// cosmos rpc provider
 pub mod rpc;
 
-/// A reference to a Cosmos chain
-#[derive(Debug)]
+/// Abstraction over a connection to a Cosmos chain
+#[derive(Debug, Clone)]
 pub struct CosmosProvider {
     domain: HyperlaneDomain,
+    canonical_asset: String,
+    grpc_client: WasmGrpcProvider,
+    rpc_client: HttpClient,
 }
 
 impl CosmosProvider {
     /// Create a reference to a Cosmos chain
-    pub fn new(domain: HyperlaneDomain) -> Self {
-        Self { domain }
+    pub fn new(
+        domain: HyperlaneDomain,
+        conf: ConnectionConf,
+        locator: Option<ContractLocator>,
+        signer: Option<Signer>,
+    ) -> ChainResult<Self> {
+        let grpc_client = WasmGrpcProvider::new(conf.clone(), locator, signer)?;
+        let rpc_client = HttpClient::builder(
+            conf.get_rpc_url()
+                .parse()
+                .map_err(Into::<HyperlaneCosmosError>::into)?,
+        )
+        // Consider supporting different compatibility modes.
+        .compat_mode(CompatMode::latest())
+        .build()
+        .map_err(Into::<HyperlaneCosmosError>::into)?;
+
+        Ok(Self {
+            domain,
+            rpc_client,
+            grpc_client,
+            canonical_asset: conf.get_canonical_asset(),
+        })
+    }
+
+    /// Get a grpc client
+    pub fn grpc(&self) -> WasmGrpcProvider {
+        self.grpc_client.clone()
```

Comment:
could this return `&WasmGrpcProvider` so we don't have to clone all the time here?